### PR TITLE
Add snippet sharing support.

### LINKED ISSUE #6: Playground: Gist support
We do not support sharing directly from the playground, since it is all client side. Adding the ability to load Github gists or Golang playground code by id should help fill in that gap.

We also might be able to add a route in angularjs like /playground/[gist-id] that automatically loads the code from the gist with that id.

### PR DESCRIPTION
**Pull request by shurcooL**
Closes #6.

This PR is a very rough start, the code needs to be cleaned up, reviewed and improved. This is my first time using AngularJS, so I may be doing things in a suboptimal way. I've already marked the things I know are poor/incorrect with TODOs.

This is to get the process started.

Known remaining tasks:
 - [x] Improve style of `share-url` text box.
       - Right now it's too small, needs a gray background, possibly not be editable.

### CODE REVIEW

File: playground/playground.go [hunk 1]
```diff
@@ -174,6 +210,38 @@ func main() {
...
+					scope.Set("showShareUrl", true)
+					// TODO: Do this better using AngularJS.
+					//       Perhaps using http://stackoverflow.com/questions/14833326/how-to-set-focus-on-input-field/18295416.
+					go func() {
+						time.Sleep(time.Millisecond)
+						dom.GetWindow().Document().GetElementByID("share-url").(*dom.HTMLInputElement).Select()
+					}()
+				})
+			})
+			// TODO: Should be sending as binary blob or is text okay?
```
Comment (shurcooL) [inline playground/playground.go]: Text is not okay, it causes last newlines to be lost/added, etc. We want byte-for-byte precise data transfer.

Comment (shurcooL) [inline playground/playground.go]: Actually, I'm not completely sure if it causes issues, it might've been due to something else.

**Comment by shurcooL** [comment]
Any suggestions on what else should be done for this to be merged?

I think it might be helpful to move the XHR request code into helper functions, since it's pretty verbose.

File: playground/playground.go [hunk 2]
```diff
@@ -13,17 +13,45 @@ import (
...
+	"honnef.co/go/js/dom"
+	"honnef.co/go/js/xhr"
 )
 
 type Line map[string]string
 
 var output []Line
 
+//const snippetStoreHost = "localhost:8080"
+const snippetStoreHost = "gotools.org:26204" // TODO: Decide on where the snippet-store server will be hosted.
```
Comment (dominikh) [inline playground/playground.go]: I would recommend hosting it at a subdomain; makes it easier to move the service to a different server without having to update the JS file.

**Comment by dominikh** [comment]
Imho, if it works, merge it and improve it in later revisions.

File: playground/playground.go (hunk 2 above)
Comment (shurcooL) [inline playground/playground.go]: I would still need a custom port, so, `snippets.gotools.org:26204`?

Comment (dominikh) [inline playground/playground.go]: For example; although I'm not sure why you need the custom port, a frontend proxy could do the right thing per-domain (proxy gotools.org:80 to your current thing, proxy snippets.gotools.org:80 to the snippet service)

Comment (shurcooL) [inline playground/playground.go]: Ok, I can do that. `snippets.gotools.org` then.

**Comment by shurcooL** [comment]
> Imho, if it works, merge it and improve it in later revisions.

I tend to agree, and I think it's basically ready now. It definitely works in my testing. I think it can be merged, but I'll wait for some feedback/review for now.

**Comment by shurcooL** [comment]
As Richard said,

> @shurcooL @dominikh @ajhager Is it okay for you guys if I leave this up to you? I don't have so much time and I would like to spend it on improving GopherJS itself. I'm sure you'll come up with a great solution.

I think it's up to us. If @dominikh and @ajhager can review and +1 this, let's go ahead and get this feature out.

Let's keep in mind one thing from the README:

> The playground is a place where users can explore GopherJS without needing to install anything, or to share code with others. Our motivation should be that a potential user could easily be turned of by a bad experience here.

I want to make sure we don't compromise the current great experience of the GopherJS Playground.

From my local testing, I do not see any more issues that need to be solved, which is why this PR has LGTM from me. If you spot any usability or other issues, let me know!

**Comment by ajhager** [comment]
LGTM! Great job on this @shurcooL :dart:

**Comment by shurcooL** [comment]
Thanks! @dominikh says it looks okay to him too, so I'll merge (i.e., deploy) this change tomorrow morning!

Yay, snippets.

**Comment by shurcooL** [comment]
Merging now. Please let me know if you spot any issues.

**Comment by shurcooL** [comment]
Well, this immediately makes one issue obvious, that I overlooked prior. The playground, actually available at http://gopherjs.org/playground/index.html, runs inside an iframe of http://gopherjs.org/play/.

But the code uses the current frame's URL to generate the shared link.

So, sharing kinda works, but the generated link has the form of:

http://gopherjs.org/playground/index.html#/Myqj0bQVev

**Pull request merged**